src/psyclone/psyir/symbols/symbol_table.py

```diff
@@ -548,21 +548,42 @@ def add(self, new_symbol, tag=None):
 
         self._symbols[key] = new_symbol
 
-    def check_for_clashes(self, other_table):
+    def check_for_clashes(self, other_table, include_arguments=True):
         '''
         Checks the symbols in the supplied table against those in
         this table. If there is a name clash that cannot be resolved by
-        renaming then a SymbolError is raised.
+        renaming then a SymbolError is raised. If `include_arguments` is
+        False then those symbols representing routine arguments in
+        `other_table` are ignored (this is used when inlining a routine
+        body into another routine).
 
         :param other_table: the table for which to check for clashes.
         :type other_table: :py:class:`psyclone.psyir.symbols.SymbolTable`
+        :param Optional[bool] include_arguments: whether to include Symbols
+            which have an ArgumentInterface in `other_table` in the checking.
 
-        :raises SymbolError: if there would be an unresolvable name clash \
+        :raises SymbolError: if there would be an unresolvable name clash
             when importing symbols from `other_table` into this table.
 
         '''
+        if include_arguments:
+            symbols_to_skip = []
+        else:
+            symbols_to_skip = other_table.argument_list[:]
+
+        try:
+            # In the case where the 'other_table' belongs to a routine,
+            # we don't want or need the symbol representing that routine.
+            rsym = other_table.lookup_with_tag("own_routine_symbol")
+            if isinstance(rsym, RoutineSymbol):
+                # We only want to skip RoutineSymbols, not DataSymbols (which
+                # we may have if we have a Fortran function).
+                symbols_to_skip.append(rsym)
+        except KeyError:
+            pass
+
```

Comment:
I agree with the proposed solution, and I am surprised that we missed this case the first time :)

However, I would suggest that method should have the "symbols_to_skip" directly as argument:
- This will make it more generic to other possible use-cases.
- The fact that inlining doesn't care about the routine and argument name belongs to the logic inside inlining. I don't think this docstring should mention it.
- In the future we could allow the transformation to skip more symbols (if psyclone can not guarantee its safety) e.g. providing a transformation option for additional symbols to skip and DSL information. And this again is transformation logic and not symbol table's.

Reply:
On a second look I see that the merge and _add_symbols_from_table methods already have a include_arguments and now I am less convinced of the change if we favour consistency. But I still think that it will be useful for the transformation to manually specify symbols to ignore the check.

What do you think?

Reply:
Your argument is quite convincing. I think we could change the interface to `merge` and `_add_symbols` so that they too take `symbols_to_skip` as argument. I'll take a look and see how much work that would be.

Reply:
It was quite straightforward and it felt good moving the logic about which symbols to exclude into the transformation. I've triggered the integration tests.